src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+            -- If we're already there, great
+            pure $ Just (cid, header)
+          else do
+            -- Otherwise, we need to do an ancestral lookup
+            case HM.lookup cid (wbhdb ^. webBlockHeaderDb) of
+              Nothing -> error "getBlockHeadersAt: Malformed WebBlockHeaderDb"
+              Just bdb -> do
+                seekAncestor bdb header (fromIntegral bh) >>= \case
+                  Just h -> do
+                    -- Sanity check, should absolutely never happen
+                    when (_blockHeight h /= bh) $ do
+                      error "getBlockHeadersAt: expected seekAncestor behaviour is broken"
+                    pure $ Just (cid, h)
+                  Nothing -> error "getBlockHeadersAt: no ancestor found!"
+        else do
+          pure Nothing
+
+-- | Make sure that the blockheight exists on chain.
+ensureBlockHeightExists :: Database -> BlockHeight -> IO ()
+ensureBlockHeightExists db bh = do
+  r <- Pact.qry db "SELECT blockheight FROM BlockHistory WHERE blockheight = ?1" [SInt (fromIntegral bh)] [RInt]
+  case r of
+    [[SInt rBH]] -> do
+      when (fromIntegral bh /= rBH) $ do
+        error "ensureBlockHeightExists: malformed query"
+    _ -> do
+      error $ "ensureBlockHeightExists: empty BlockHistory: height=" ++ show bh
+
+-- | Get the earliest blockheight on chain.
+getEarliestBlockHeight :: Database -> IO BlockHeight
+getEarliestBlockHeight db = do
+  r <- Pact.qry db "SELECT blockheight FROM BlockHistory ORDER BY blockheight ASC LIMIT 1" [] [RInt]
+  case r of
+    [[SInt bh]] -> do
+      pure (fromIntegral bh)
+    _ -> do
+      error "getEarliestBlockHeight: no earliest blockheight"
+
+-- | Get the latest blockheight on chain.
+getLatestBlockHeight :: Database -> IO BlockHeight
+getLatestBlockHeight db = do
+  r <- Pact.qry db "SELECT blockheight FROM BlockHistory ORDER BY blockheight DESC LIMIT 1" [] [RInt]
+  case r of
+    [[SInt bh]] -> do
+      pure (fromIntegral bh)
+    _ -> do
+      error "getLatestBlockHeight: no latest blockheight"
+
+-- | Wrapper around 'withSqliteDb' that adds the chainId label to the logger
+--   and sets resetDb to False.
+withChainDb :: (Logger logger)
+  => ChainId
+  -> logger
+  -> FilePath
+  -> (SQLiteEnv -> IO x)
+  -> IO x
+withChainDb cid logger' path f = do
+  let logger = addLabel ("chainId", chainIdToText cid) logger'
+  let resetDb = False
+  withSqliteDb cid logger path resetDb f
+
+-- | Like 'TargetBlockHeight', but supports multiple targets in the non-'Latest'
+--   case.
+data BlockHeightTargets
+  = LatestAll
+  | TargetAll (Set BlockHeight)
+
+-- | Resolve the requested targets across all chains.
+--
+--   In the 'Latest' case, not all chains may be at the same tip,
+--   so we have to get the highest block that they each have in their
+--   history. This operates under the assumption that all chains
+--   are moving along at roughly the same pace.
+--
+--   Returns an allowlist of blockheights for each chain.
```

Comment:
allowlist isn't the correct term here. i'm not sure what a good word would be. but essentially, the set associated with each ChainId in the map is every BlockHeight we determined is actually  available for that Chain.